modules/src/events.rs

```diff
@@ -0,0 +1,250 @@
+                vals.push(ev);
+            }
+        }
+        _ => {}
+    }
+    Ok(vals)
+}
+
+pub fn build_event(object: RawObject) -> Result<IBCEvent, BoxError> {
+    match object.action.as_str() {
+        "create_client" => Ok(IBCEvent::from(ClientEvents::CreateClient::try_from(
+            object,
+        )?)),
+        "update_client" => Ok(IBCEvent::from(ClientEvents::UpdateClient::try_from(
+            object,
+        )?)),
+
+        "connection_open_init" => Ok(IBCEvent::from(ConnectionEvents::OpenInit::try_from(
+            object,
+        )?)),
+        "connection_open_try" => Ok(IBCEvent::from(ConnectionEvents::OpenTry::try_from(object)?)),
+        "connection_open_ack" => Ok(IBCEvent::from(ConnectionEvents::OpenAck::try_from(object)?)),
+        "connection_open_confirm" => Ok(IBCEvent::from(ConnectionEvents::OpenConfirm::try_from(
+            object,
+        )?)),
+
+        "channel_open_init" => Ok(IBCEvent::from(ChannelEvents::OpenInit::try_from(object)?)),
+        "channel_open_try" => Ok(IBCEvent::from(ChannelEvents::OpenTry::try_from(object)?)),
+        "channel_open_ack" => Ok(IBCEvent::from(ChannelEvents::OpenAck::try_from(object)?)),
+        "channel_open_confirm" => Ok(IBCEvent::from(ChannelEvents::OpenConfirm::try_from(
+            object,
+        )?)),
+        "channel_close_init" => Ok(IBCEvent::from(ChannelEvents::CloseInit::try_from(object)?)),
+        "channel_close_confirm" => Ok(IBCEvent::from(ChannelEvents::CloseConfirm::try_from(
+            object,
+        )?)),
+
+        // send_packet
+        "transfer" => Ok(IBCEvent::from(ChannelEvents::SendPacket::try_from(object)?)),
+        // recv_packet
+        "ics04/opaque" => Ok(IBCEvent::from(ChannelEvents::ReceivePacket::try_from(
+            object,
+        )?)),
+        // acknowledge_packet
+        // needs these changes in cosmos-sdk
+        //        --- a/x/ibc/04-channel/types/msgs.go
+        //        +++ b/x/ibc/04-channel/types/msgs.go
+        //    @@ -511,5 +511,5 @@ func (msg MsgAcknowledgement) GetSigners() []sdk.AccAddress {
+        //
+        //        // Type implements sdk.Msg
+        //        func (msg MsgAcknowledgement) Type() string {
+        //            -       return "ics04/opaque"
+        //            +       return "ics04/acknowledge"
+        //        }
+        "ics04/acknowledge" => Ok(IBCEvent::from(ChannelEvents::AcknowledgePacket::try_from(
+            object,
+        )?)),
+        //timeout_packet
+        "ics04/timeout" => Ok(IBCEvent::from(ChannelEvents::TimeoutPacket::try_from(
+            object,
+        )?)),
+
+        // TODO not clear what the message.action for this is
+        "cleanup_packet" => Ok(IBCEvent::from(ChannelEvents::CleanupPacket::try_from(
+            object,
+        )?)),
+
+        _ => Err("Incorrect Event Type".into()),
+    }
+}
+
+pub fn extract_block_height(result: &ResultEvent) -> Result<block::Height, BoxError> {
```

Comment:
Can parameter be `TMEventData::EventDataNewBlock` instead of `ResultEvent`?

Reply:
Oh, actually this is not used anymore. will remove.